Update `targetSdkVersion` (from `22` to `31`)

ParaskP7 opened the pull request:
This PR upgrades `targetSdkVersion` to `31`:

1. Mainly, for the `example` module (currently on `targetSdkVersion 22`),
2. But also, for the `plugins:woocommerce` module (currently on `targetSdkVersion 30`).

With this update, and the fact that `targetSdkVersion` will get aligned too, all modules are now pointing to the exact same `minSdkVersion`, `compileSdkVersion` and `targetSdkVersion`.

-----

Warning (⚠️): This update is necessary as it unblocks updating the common `minSdkVersion` to `24` (see [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2564)). This update is currently not possible because the `example` module `targetSdkVersion` is pointing to `22`, and as such updating is not allowed as the `targetSdkVersion` value cannot be lower than the `mindSdkVersion` value.

-----

As part of this `targetSdkVersion = 31` upgrade, the more specifically the final 6f99c4d14e72169484412b77d9d87752b007601c commit, the below changes got applied in order to fix any build/test failures that this change brought-up:

1. The `android:exported="true"` attribute was added to to the `main` related `AndroidManifest.xml` configuration for the `MainExampleActivity` related `activity` element.

This was done because otherwise assemble is failing with the below failure due to the fact that on apps targeting Android 12 this attribute needs to be explicitly specified, see below:

```
android:exported needs to be explicitly specified for element
<activity#org.wordpress.android.fluxc.example.MainExampleActivity>.
Apps targeting Android 12 and higher are required to specify an
explicit value for `android:exported` when the corresponding
component has an intent filter defined."
```

See [manifest/activity-element#exported](https://developer.android.com/guide/topics/manifest/activity-element#exported) for details.

2. The `RetryOnRedirectBasicNetworkTest` unit test suite got ignored and has now been (temporarily) excluded from the test suites.

This was done because otherwise CI would fail the test check and it would not be possible to merge this update. The actual reason why this `ClassNotFoundException` started occurring is still unknown and will be investigated as part of another PR, which will make sure to re-instantiate this `RetryOnRedirectBasicNetworkTest` test, see below:

```
Caused by: java.lang.ClassNotFoundException:
org.apache.http.StatusLine
```

3. The `android:usesCleartextTraffic="true"` attribute was added to the `main` related `AndroidManifest.xml` configuration for the `application` element.

This was done because otherwise some of the connected test are failing with the below exception due to the fact that these tests still depend on `http` instead of `https`, see below:

```
Caused by: java.net.UnknownServiceException: CLEARTEXT communication
   to do.wpmt.co not permitted by network security policy
```

-----

In addition to the above, as part of this `targetSdkVersion = 31` upgrade, the more specifically the final 6f99c4d14e72169484412b77d9d87752b007601c commit, the below configuration change got applied in order to overcome the permission issues that this change brought-up:

Both, the `android.permission.INTERNET` normal permission and the `android.permission.READ_EXTERNAL_STORAGE` dangerous permission, were moved from the `fluxc` library  module and into the `example` app module. This was done because it is not necessary, or at least no longer necessary, for the `fluxc` library module to have such a permission.

PS: This `android.permission.READ_EXTERNAL_STORAGE` dangerous permission is required for uploading media files. Without having this permission a user wouldn't be able to navigate to `FluxC Example` app's `App Info` section and manually enable this permission from within the `App Permissions` section. This is necessary because otherwise a user wouldn't be able to test the `MEDIA -> UPLOAD` functionality from within the app itself as it would otherwise throws the following error, see below:

`Upload error: FS_READ_PERMISSION_DENIED`

FYI: This whole `Media` related functionality and as such its dangerous permission was introduced as part of the below `Media Store` related PR:

- Main PR: #122
- Extra PR: #163

However, since then, a lot has changed, including all this permission related functionality, which got completely removed, leaving only this dangerous permission configuration available, most probably, just for manual testing purposes.

-----

PS.1: @oguzkocer and @maxme I am adding you as main reviewers to take a look at that `example` module related `targetSdkVersion 31` and its `android.permission.READ_EXTERNAL_STORAGE` related change as I am not sure if I am missing something here. I've read all the permission related conversations and file changes in this #122 PR but I can't be to 100% that I'm still not missing something. All the testing I need verifies that we can indeed now upgrade the `targetSdkVersion` from `22` to `31` and manually enabled the `Photos and videos` permission when doing manual testing for `Media`. For Unit and Connected tests, I am seeing all of them passing both, locally and on CI (not true for `Connected Tests`, see [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2563#issuecomment-1305720734)), thus I feel that this will not cause us such issues. I am mainly referring to this build related `targetSdkVersion` comment below, which I think no longer applies, wdyt:

```
Keep the targetSdkVersion 22 so we don't need to grant runtime permissions to the tests and the example app
An alternative would be granting the permissions via adb before running the test, like here:
https://afterecho.uk/blog/granting-marshmallow-permissions-for-testing-flavoured-builds.html
```

PS.2 @malinajirka I am adding you as an extra reviewer to help identify any potential issues with that upgrade, especially since you also have lots of experience with `FluxC` were working with `WPAndroid` for few years and are now working with `WCAndroid` for more than a year. As such, your help here would be very valuable.

-----

### Testing instructions

- Quickly smoke test the `FluxC Example` app and see if it works as expected.
- Try and focus on permission testing, especially on `READ_EXTERNAL_STORAGE` and the `Media` related functionality of the `FluxC Example` app. See how the app behaves with and without having this permission enabled.

ParaskP7 committed:
Build: Extract common min sdk version to settings for all modules

This will now make sure that from now on all module will be pointing to
the same 'minSdkVersion' versions, making such upgrades much more
straightforward as that will apply to all modules. This will alleviate
the risk of forgetting to update a specific module, or set of modules
due to an oversight.

ParaskP7 committed:
Build: Extract common compile sdk version to settings for all modules

This will now make sure that from now on all module will be pointing to
the same 'compileSdkVersion' versions, making such upgrades much more
straightforward as that will apply to all modules. This will alleviate
the risk of forgetting to update a specific module, or set of modules
due to an oversight.

ParaskP7 committed:
Build: Extract common target sdk version to settings for fluxc module

This will now allow all modules to be pointing to the same common
'targetSdkVersion' versions, making such upgrades much more
straightforward as that should apply to all modules. This will alleviate
the risk of forgetting to update a specific module, or set of modules
due to an oversight.

ParaskP7 committed:
Build: Update target sdk version to 31 for plugins woocommerce module

Warnings Message: "Google Play requires that apps target API level 31 or
higher."

This 'plugins:woocommerce' library was still targeting SDK version 30
instead of 31. As per the Google Play requirement and the fact that
the Woo Commerce app itself is already targeting 'targetSdkVersion = 31'
for some time now (about a year, see WCAndroid PR link below), this
change updates this module's target SDK version to 31.

PS: This update and the fact that this is a 'targetSdkVersion' update on
a library module should not cause any problems or need any additional
testing.

WCAndroid PR Link: https://github.com/woocommerce/woocommerce-android/
pull/5348

ParaskP7 committed:
Build: Update target sdk version to 31 for example module

This update is necessary as it unblocks updating the
common 'minSdkVersion' to '24'. This update is currently not possible
because the 'example' module 'targetSdkVersion' is pointing to '22',
and as such updating is not allowed as the 'targetSdkVersion'
value cannot be lower than the 'mindSdkVersion' value.

------------------------------------------------------------------------

In order to make it possible for the 'example' module to have its
'targetSdkVersion' set to '31' the below changes were necessary:

1) The 'android:exported="true"' attribute was added to to the 'main'
   related 'AndroidManifest.xml' configuration for the
   'MainExampleActivity' related 'activity' element.

   This was done because otherwise assemble is failing with the below
   failure due to the fact that on apps targeting Android 12 this
   attribute needs to be explicitly specified, see below:

   "android:exported needs to be explicitly specified for element
   <activity#org.wordpress.android.fluxc.example.MainExampleActivity>.
   Apps targeting Android 12 and higher are required to specify an
   explicit value for `android:exported` when the corresponding
   component has an intent filter defined."

   See https://developer.android.com/guide/topics/manifest/
   activity-element#exported for details.
2) The 'RetryOnRedirectBasicNetworkTest' unit test suite got ignored and
   has now been (temporarily) excluded from the test suites.

   This was done because otherwise CI would fail the test check and it
   would not be possible to merge this update. The actual reason why
   this 'ClassNotFoundException' started occurring is still unknown and
   will be investigated as part of another PR, which will make sure to
   re-instantiate this 'RetryOnRedirectBasicNetworkTest' test.

   "Caused by: java.lang.ClassNotFoundException:
   org.apache.http.StatusLine"
3) The 'android:usesCleartextTraffic="true"' attribute was added to the
   'main' related 'AndroidManifest.xml' configuration for the
   'application' element.

   This was done because otherwise some of the connected test are
   failing with the below exception due to the fact that these tests
   still depend on 'http' instead of 'https', see below:

   "Caused by: java.net.UnknownServiceException: CLEARTEXT communication
   to do.wpmt.co not permitted by network security policy"

------------------------------------------------------------------------

In addition to the above, and in order to make it possible for the
'example' module to have its 'targetSdkVersion' set to '31' the below
configuration got updated.

Both, the 'android.permission.INTERNET' normal permission and the
'android.permission.READ_EXTERNAL_STORAGE' dangerous permission, were
moved from the 'fluxc' library module and into the 'example' app module.
This was done because it is not necessary, or at least no longer
necessary, for the 'fluxc' library module to have such a permission.

PS: This 'android.permission.READ_EXTERNAL_STORAGE' dangerous permission
is required for uploading media files. Without having this permission a
user wouldn't be able to navigate to 'FluxC Example' app's 'App Info'
section and manually enable this permission from within the
'App Permissions' section. This is necessary because otherwise a user
wouldn't be able to test the 'MEDIA -> UPLOAD' functionality from within
 the app itself as it would otherwise throws the following error:

"Upload error: FS_READ_PERMISSION_DENIED"

------------------------------------------------------------------------

FYI: This whole 'Media' related functionality and as such its
dangerous permission was introduced as part of the below 'Media Store'
related PR:

Main PR: https://github.com/wordpress-mobile/WordPress-FluxC-Android
/pull/122
Extra PR: https://github.com/wordpress-mobile/WordPress-FluxC-Android
/pull/163

However, since then, a lot has changed, including all this permission
related functionality, which got completely removed, leaving only this
dangerous permission configuration available, most probably, just for
manual testing purposes.

ParaskP7 commented:
👋 @oguzkocer !

About f410f33beeb358edfb10437d8d444f8d8c6ac347 and the fact that the `Connected Tests` are still disabled on `FluxC` as we are still not publishing the GitHub status. Do we have an update? 🤔

It would have helped to get a 👍 from CI for the `Connected Tests` check in this PR.

[Connected Tests](https://buildkite.com/automattic/wordpress-fluxc-android/builds/1812#0184528b-d6c2-4556-8a88-31b5ba70e4dc): Status ❌
[Firebase Failed Tests](https://buildkite.com/organizations/automattic/analytics/suites/fluxc-android-connected-tests/runs/5043d1c3-cf35-4fbc-948d-ef459029f1c7): 3
1. `ReleaseStack_MediaTestWPCom.testFetchMediaList()`
2. `ReleaseStack_MediaTestWPCom.testFetchMedia()`
3. `ReleaseStack_UploadTest.testUploadMediaError()`

PS: Locally, I verified that tests were passing, including the above 3 tests that failed on Firebase, even when locally the app doesn't have any permissions enabled, well, give or take this `ReleaseStack_PluginTestJetpack.testInstallAndDeleteSitePlugin` test, which was constantly failing for me, but I think this test is actually unrelated.

malinajirka commented:
Thanks for working on this @ParaskP7! I've tested the example app and WCAndroid and I haven't noticed any issues. LGTM :)

ParaskP7 commented:
👋 @malinajirka !

> Thanks for working on this @ParaskP7! I've tested the example app and WCAndroid and I haven't noticed any issues. LGTM :)

Thank you so much for the review and testing the `example` app, this is much appreciated! 🙇 

Did you maybe also tried to trigger the connected test on your site and see if those are working as expected as well? 🤔

On my side, locally, they are working fine, but on CI I see few media related test failures (see [above](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2563#issuecomment-1305720734)), which might or might not be related, as I am seeing other such media tests passing anyway, but I can't be for sure.

Thus, it would help if we at least get another verification that the connected tests are working locally, especially since the Buildkite `Connected Tests` CI check is still disabled for this repo.

malinajirka commented:
This test keeps failing for me - testFetchAuthOptionsForUserWithPassword - but it fails even on trunk, so not related to this PR.

ParaskP7 commented:
Thank you a lot for running through the connected test @malinajirka and sharing your results, much appreciated! 🙇 

> This test keeps failing for me - testFetchAuthOptionsForUserWithPassword - but it fails even on trunk, so not related to this PR.

Yes, this test seems to not be related to this change, thanks for trying it on `trunk` as well. 🥇

ParaskP7 commented:
👋 @oguzkocer !

I wonder whether this `Media` related test failures could be somehow related to [this](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2391) update `TEST_WPCOM_IMAGE_IDS_TEST1` work you did back then, in the sense that maybe those are not related to the `targetSdkChange`, but that instead we need to update our encrypted secrets? 🤔

I am thinking that because it feels a bit off to me to have those specific `Media` tests failing and not all of them, especially those that are flowing through and dispatching the `MediaActionBuilder.newUploadMediaAction(payload)` event, about `16` of them, with only `3` of them failing. 🤔 🤔

In addition to that, the fact that they are not failing locally seems a bit off too. 🤔 🤔 🤔

oguzkocer commented:
@ParaskP7 It's possible, but I haven't touched the site contents since that fix, so I don't know why it'd start to fail. Unfortunately, for the moment, we are accepting that some FluxC connected tests are failing. I've put in a lot of effort to fixing them a while back, but we still had issues in Firebase (devices going offline before the success state is reported - not sure if it's due to them or us).

The way the tests are currently setup is prone to failure and it's been a losing fight to address them as is. So, we want to figure out a way to make the tests more robust. However, that's currently on hold because my understanding is that there is some possibility that we might not need to work with Firebase in the future and we'd like to wait until that situation is resolved first.

ParaskP7 commented:
👋 @oguzkocer and thank you for your comment! 🙇 

> @ParaskP7 It's possible, but I haven't touched the site contents since that fix, so I don't know why it'd start to fail. 

👍 

> Unfortunately, for the moment, we are accepting that some FluxC connected tests are failing. I've put in a lot of effort to fixing them a while back, but we still had issues in Firebase (devices going offline before the success state is reported - not sure if it's due to them or us).

Yea, I know... 😞 

> The way the tests are currently setup is prone to failure and it's been a losing fight to address them as is. So, we want to figure out a way to make the tests more robust.

Yeap, I remember this discussion. 👍 

> However, that's currently on hold because my understanding is that there is some possibility that we might not need to work with Firebase in the future and we'd like to wait until that situation is resolved first.

Oh, I didn't know about us possibly stopping using Firebase for that matter, thanks for letting me know. 👍

ParaskP7 commented:
👋 @oguzkocer !

As per our above conversation and if you've also run the connected tests on yours side and see them passing locally, at least the `Media` related ones, do you feel that we can progress with merging this PR to `trunk` and deal with these connected tests later on, that is when we figure out what to do with them in general, Firebase and all? 🤔

I personally vote in favor of moving forward, rather than keeping `targetSdkVersion` at `22`, even if that's only for the `example` app. Truth being told, we do use that app quick often and it hasn't been maintained properly that way.

I would love for us to start removing the blockers on these kind of maintenance, and that's a good first step, even if it means that some `Media` related tests are to be ignored later on, or fixed properly, or converted to another type of test, and to be honest, there are not lots of test like that failing, and all of them on CI only.

Wdyt? 🤔

oguzkocer commented on settings.gradle:
```diff
@@ -86,3 +86,9 @@ include ':fluxc',
         ':example',
         ':tests:api'
+
+gradle.ext {
+    minSdkVersion = 21
+    compileSdkVersion = 31
+    targetSdkVersion = 31
+}
```
The convention is that, if these values are not used in `settings.gradle` file, they should be in `build.gradle` instead.

ParaskP7 replied:
👋 @oguzkocer and thank you for your comment! 💯

I was torn as well in making this decision, that is, choosing the [WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/build.gradle#L20-L22) way, or the [WCAndroid](https://github.com/woocommerce/woocommerce-android/blob/trunk/settings.gradle#L85-L89) way of defining such SDK version values.

I ended up choosing the [WCAndroid](https://github.com/woocommerce/woocommerce-android/blob/trunk/settings.gradle#L85-L89) way because I find it more appropriate to differentiate the SDK updates to the normal library dependency versions, with the plugin dependency versions also being separate.

I don't have a strong opinion in where to have those SDK versions, `settings.gradle` or `build.gradle` and your reasoning is valid to me. I just wanted to apply this change in such a way that the SDK version are separate to the library dependency versions as am not a fan of having it the way [WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/build.gradle#L20-L22) does.

Wdyt? 🤔

PS: Having those values within `settings.gradle` or within the root level `build.gradle` still makes them unused, they are only used on the module level `build.gradle`, but I am not sure if you meant that they should be in a file named as `build.gradle` as a convention. Actually, in the past I would have also extracted them in separate Gradle files, like `android.gradle` and apply that file where needed to get those version related values, with Kotlin Gradle DSL, one could also use something like a [convention plugin](https://github.com/android/nowinandroid/blob/main/build-logic/convention/src/main/kotlin/AndroidApplicationConventionPlugin.kt#L36) (nowadays that `buildSrc` is no longer a thing).

-----

No matter, just let me know, and if you feel strong to have them in `build.gradle` I'll make this change here, as well as on such [Utils](https://github.com/wordpress-mobile/WordPress-Utils-Android/pull/111) and [Login](https://github.com/wordpress-mobile/WordPress-Login-Flow-Android/pull/103) related PRs of mine. 💯

oguzkocer replied:
The convention is to add these to the `build.gradle`, because when we do that, we define these as [project extension](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html), also referred to as `project.ext`, or the DSL shorthand `ext`. When we add it to the `gradle.ext`, it becomes an extension of the [global Gradle object](https://docs.gradle.org/current/javadoc/org/gradle/api/invocation/Gradle.html) which is really more of a hack. The only reason we have done that for `includedBuild`s was that there was no other way to share the value between `settings.gradle` & `build.gradle` files. I think now there are other ways, such as version catalog.

If you don't like to put them in the `build.gradle` file, you can create a separate `.gradle` file to organize the build files. It's also possible to split the `project.ext` definitions to multiple categories within the `.gradle` files. It's also possible to use a local version catalog. So, the more important point I am trying to make is that we don't want to define them as `gradle.ext` because it's not necessary and it puts it outside of the project lifecycle.

I hope that makes sense.

ParaskP7 replied:
Yes, it makes sense and thanks for the extra clarification on all that @oguzkocer ! 🙇 

> The convention is to add these to the build.gradle, because when we do that, we define these as [project extension](https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html), also referred to as project.ext, or the DSL shorthand ext. When we add it to the gradle.ext, it becomes an extension of the [global Gradle object](https://docs.gradle.org/current/javadoc/org/gradle/api/invocation/Gradle.html) which is really more of a hack. The only reason we have done that for includedBuilds was that there was no other way to share the value between settings.gradle & build.gradle files. I think now there are other ways, such as version catalog.

👍 

> If you don't like to put them in the build.gradle file, you can create a separate .gradle file to organize the build files. It's also possible to split the project.ext definitions to multiple categories within the .gradle files. It's also possible to use a local version catalog. So, the more important point I am trying to make is that we don't want to define them as gradle.ext because it's not necessary and it puts it outside of the project lifecycle.

Let me think about that and I'll come back with an alternative (probably on Monday) so we can discuss and agree on before I proceed with any change, or merging this PR for that matter. 🙏 

Once more, thank you Oguz! 🥇

ParaskP7 replied:
👋 @oguzkocer !

> Let me think about that and I'll come back with an alternative (probably on Monday) so we can discuss and agree on before I proceed with any change, or merging this PR for that matter. 🙏

I ended-up thinking that it is better, at least at this point of time, to go with the `project.ext` alternative, that is, instead of the `gradle.ext`, as per your recommendation. As such, I recommend moving the below `settings.gradle` + `gradle.ext` related configuration:

`settings.gradle`
```
gradle.ext {
    minSdkVersion = 21
    compileSdkVersion = 31
    targetSdkVersion = 31
}
```

Into it's own separate such `project.ext` configuration on the root level `build.gradle`:

`build.gradle` (root)

```
ext {
    minSdkVersion = 21
    compileSdkVersion = 31
    targetSdkVersion = 31
}
```

This way, this will become a first good step in splitting these SDK versions from the rest of the dependency versions, but still utilizing the `rootProject.xyzSdkVersion` instead of the `gradle.ext.xyzSdkVersion`.

Do you also agree with that? 😊

oguzkocer replied:
Yeap, that sounds good to me. Thanks!

ParaskP7 replied:
Great, thank you so much for the confirmation on that Oguz, I'll now proceed with this change here and everywhere else needed. 🥇

oguzkocer reviewed (approved):
@ParaskP7 All media tests are passing for me locally. I left a comment about general Gradle convention, but otherwise it looks good to me 👍

ParaskP7 committed:
Build: Replace gradle ext with project ext for sdk versions declaration

For more info on that check PR comment and relevant discussion below:

https://github.com/wordpress-mobile/WordPress-FluxC-Android/
pull/2563#discussion_r1020079074

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-FluxC-Android into build/update-target-sdk-version-to-31

ParaskP7 committed:
Merge branch 'trunk' of github.com:wordpress-mobile/WordPress-FluxC-Android into build/update-target-sdk-version-to-31
